azure-batch/azure/batch/custom/patch.py

```diff
@@ -0,0 +1,287 @@
+    omit tasks that were successfully added on the first attempt. The
+    maximum lifetime of a task from addition to completion is 7 days. If a
+    task has not completed within 7 days of being added it will be
+    terminated by the Batch service and left in whatever state it was in at
+    that time.
+
+    :param job_id: The ID of the job to which the task collection is to be
+        added.
+    :type job_id: str
+    :param value: The collection of tasks to add. The total serialized
+        size of this collection must be less than 4MB. If it is greater than
+        4MB (for example if each task has 100's of resource files or
+        environment variables), the request will fail with code
+        'RequestBodyTooLarge' and should be retried again with fewer tasks.
+    :type value: list of :class:`TaskAddParameter
+        <azure.batch.models.TaskAddParameter>`
+    :param task_add_collection_options: Additional parameters for the
+        operation
+    :type task_add_collection_options: :class:`TaskAddCollectionOptions
+        <azure.batch.models.TaskAddCollectionOptions>`
+    :param dict custom_headers: headers that will be added to the request
+    :param bool raw: returns the direct response alongside the
+        deserialized response
+    :param int threads: number of threads to use in parallel when adding tasks. If specified
+        and greater than 0, will start additional threads to submit requests and wait for them to finish.
+        Otherwise will submit add_collection requests sequentially on main thread
+    :return: :class:`TaskAddCollectionResult
+        <azure.batch.models.TaskAddCollectionResult>` or
+        :class:`ClientRawResponse<msrest.pipeline.ClientRawResponse>` if
+        raw=true
+    :rtype: :class:`TaskAddCollectionResult
+        <azure.batch.models.TaskAddCollectionResult>` or
+        :class:`ClientRawResponse<msrest.pipeline.ClientRawResponse>`
+    :raises:
+        :class:`BatchErrorException<azure.batch.models.BatchErrorException>`
+    """
+
+    results_queue = collections.deque()  # deque operations(append/pop) are thread-safe
+    task_workflow_manager = _TaskWorkflowManager(
+        client,
+        job_id,
+        value,
+        task_add_collection_options,
+        custom_headers,
+        raw,
+        **operation_config)
+
+    # multi-threaded behavior
+    if threads:
+        if threads < 0:
+            raise ValueError("Threads must be positive or 0")
+
+        active_threads = []
+        for i in range(threads):
+            active_threads.append(threading.Thread(
+                target=task_workflow_manager.task_collection_thread_handler,
+                args=(results_queue,)))
+            active_threads[-1].start()
+        for thread in active_threads:
+            thread.join()
+    # single-threaded behavior
+    else:
+        task_workflow_manager.task_collection_thread_handler(results_queue)
+
+    submitted_tasks = _handle_output(results_queue)
```

Comment:
I was wondering about moving this line into the else statement below - but I see it can be used to raise any unexpected/unrecognized exceptions that occurred during requests.
I guess it comes down to which exception you wish to take priority in raising to the user - either an error stored in task_workflow_manager.error, and a request exception (if both happened to have occurred).
Your call. :)

Reply:
I feel like if there was an unexpected exception it should be given priority as we would want to explain why we didn't finish serving the request. 
Since we are losing the original stack trace from re-raising it anyways, one idea would be to add it to the custom error raised that way the only exception need to be caught would be CreateTasksErrorException.  Then we could effectively communicate client failures + tasks not submitted + earlyTerminationErrors in one blob..

Thoughts?

Reply:
@matthchr @xingwu1 as well

Reply:
Given that you already have the error structure in place to report back to the user which tasks have/have not been submitted - I think it would make sense to reuse this for other the 'Unrecognized error' scenario... especially something retryable like an intermittent connection error